dlib/gui_widgets/widgets.cpp

```diff
@@ -6059,6 +6062,15 @@ namespace dlib
         overlay_rects[selected_rect].parts[part_name] = c1;
         parent.invalidate_rectangle(rect); 
 
+        // trz
+        // I do this here so that it advances when parts
+        // are added from the context menu too
+        // This also allows to restart the sequence from
+        // a random point using the context menu
+        int part_idx = std::distance(part_names.begin(), part_names.find(part_name));
```

Comment:
This is wonky.  It would be clearer to save the string part name and use that to know what part is next.  Otherwise there are all kinds of problems, like what happens when parts are removed from the list and now the indexes don't make sense.

Reply:
If they start to edit parts then everything is wonky. I didn't know it was possible, is it?

Saving the part name does not help as I may end up adding a removed part or I should check if the part is still there. And which is the part after the one that is no longer in the list?
Here I'm "iterating" over a list so I think it's clearer to use an index rather than an element and do a lookup to get back the index every time I need to move to the next. The semantic is that of an index, I wouldn't mask it to handle(?) some borderline cases.

What I can do is to reset the index to zero if the index is out of bounds. Or I could reset the index to zero every time the part list is modified.

But I cannot find the parts edit code, is there or maybe we are worrying about something that may never happen?

Reply:
Yes, it is.  Not though imglab, but you have to look at the interface for the objects you are editing.  Those objects allow this, you can't just do things in some functions that will cause subsequent valid uses of the object to explode.  This is all the more important since dlib is a widely used library and who knows how many uses of these objects exist in other applications.

Yes, you have to check for everything that might happen.  Or to say it another way, you can never put the object into an invalid state.  Also, the entire API for all the objects in dlib is documented in great detail (see also http://dlib.net/howto_contribute.html#1).  These functions will say things like "Calling this function removes a part.".  It will never say something like "calling this function probably removes a part, or makes your program crash depending on how the users has been moving the mouse." because that would be a terrible contract for the function.  Anyway, my point is that you can read the documented interface for these objects to find out exactly what constitutes valid behavior and "valid vs. invalid state".  

As for the index, if the parts list changes then an integer index is meaningless.  Another thing to consider is that any time the parts list changes you should probably seek back to the start of the list.  In any case, an integer index is not inherently bad, but the code you have here is constantly dealing in the string names anyway.  You might as well skip the integer part and just use the string since the code would be simpler and detection of unexpected changes more durable.